File: fixtures/scale/dart/large/lib/ciphers/playfair/playfair_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'playfair_digraphs.dart';
import 'playfair_key.dart';
import 'playfair_rules.dart';

/// Digraph substitution over a 5x5 keyword square.
class PlayfairCipher implements Cipher {
  final String _square;

  PlayfairCipher(PlayfairKey key) : _square = key.square();

  @override
  String name() => 'playfair';

  @override
  String encrypt(String plaintext) =>
      transformPairs(_square, splitDigraphs(cleanAlphabet(plaintext)), 1);

  @override
  String decrypt(String ciphertext) =>
      transformPairs(_square, splitDigraphs(cleanAlphabet(ciphertext)), 4);
}

File: fixtures/scale/scala/large/src/cryptobench/ciphers/playfair/PlayfairCipher.scala
package cryptobench.ciphers.playfair

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Digraph substitution over a 5x5 keyword square. */
final class PlayfairCipher(key: PlayfairKey) extends Cipher {
  private val square = key.square

  override def name: String = "playfair"

  override def encrypt(plaintext: String): String =
    PlayfairRules.transform(square, PlayfairDigraphs.split(Alphabet.clean(plaintext)), 1)

  override def decrypt(ciphertext: String): String =
    PlayfairRules.transform(square, PlayfairDigraphs.split(Alphabet.clean(ciphertext)), 4)
}
